### PR TITLE
Update exawind packages

### PR DESCRIPTION
**Pull request by psakievich**
Upstream some of our package changes from developer to mainline spack

- [x] exawind
- [x] amr-wind
- [x] nalu-wind
- [x] openfast
- [x] hypre

### CODE REVIEW

**Comment by eugeneswalker** [comment]
This is awesome! With these changes, what specs are most useful for CUDA and ROCm builds of ExaWind?

**Comment by psakievich** [comment]
@eugeneswalker not sure I follow what you mean by most useful for cuda and rocm builds.  This PR will let you build the stack for either case by adding `+rocm` or `+cuda`.  We aren't allowing hybrid architecture splits (amr-wind on device nalu-wind on host) as a public facing capability right now so it simplifies the flags for device quite a bit.

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by eugeneswalker** [comment]
> @eugeneswalker not sure I follow what you mean by most useful for cuda and rocm builds. This PR will let you build the stack for either case by adding `+rocm` or `+cuda`. We aren't allowing hybrid architecture splits (amr-wind on device nalu-wind on host) as a public facing capability right now so it simplifies the flags for device quite a bit.

What I mean is, since you are intimately familiar w the stack and its use, what variants would you turn on for (1)  a build intended to use nvidia gpu and (2) a build intended for amd gpu. I imagine different configurations might have different use case so just seeking your recommendation on the variants you’d advise to build w the maximize potential of it being useful for cases (1) and (2). If that makes sense :)

**Comment by psakievich** [comment]
Ah I see. I am hoping to have the defaults do that automatically so you just need to specify the minimum spec.  I don't think I have every corner case figured out, but we will just have to iterate over time.

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 1]
```diff
@@ -6,35 +6,118 @@
...
-    depends_on("openfast+cxx+shared@2.6.0:")
+    variant("openfast", default=False, description="Enable OpenFAST integration")
+    variant("hypre", default=True, description="Enable hypre solver")
+    variant("stk_simd", default=False, description="Enable SIMD in STK")
+    variant("umpire", default=False, description="Enable Umpire")
+    variant("tiny_profile", default=False, description="Turn on AMR-wind with tiny profile")
+    variant("sycl", default=False, description="Enable SYCL backend for AMR-Wind")
+    variant("gpu-aware-mpi", default=False, description="gpu-aware-mpi")
+
+    conflicts("+hypre", when="+sycl")
```
Comment (jrood-nrel) [inline var/spack/repos/builtin/packages/exawind/package.py]: hypre can build with `+sycl` so I think we can remove this.

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 2]
```diff
@@ -6,35 +6,118 @@
...
+    depends_on("amr-wind+hypre", when="+hypre")
+    depends_on("amr-wind~hypre", when="~hypre")
+    depends_on("nalu-wind~hypre", when="~hypre")
+    depends_on("trilinos+ninja", when="+ninja")
+    depends_on("nalu-wind+ninja", when="+ninja")
+    depends_on("amr-wind+ninja", when="+ninja")
+    depends_on("amr-wind+sycl", when="+sycl")
+    # not required but added so these get picked up as a
+    # direct dependency when creating snapshots
+    depends_on("trilinos")
```
Comment (jrood-nrel) [inline var/spack/repos/builtin/packages/exawind/package.py]: Should we remove these direct dependencies since they aren't necessary?

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 3]
```diff
@@ -6,35 +6,118 @@
...
+            args.append("-DCMAKE_CXX_COMPILER={0}".format(self.spec["hip"].hipcc))
+            args.append("-DCMAKE_HIP_ARCHITECTURES=" + ";".join(str(x) for x in targets))
+            args.append("-DAMDGPU_TARGETS=" + ";".join(str(x) for x in targets))
+            args.append("-DGPU_TARGETS=" + ";".join(str(x) for x in targets))
+
+        if spec.satisfies("^amr-wind+hdf5"):
+            args.append(self.define("H5Z_ZFP_USE_STATIC_LIBS", True))
+
+        if spec.satisfies("^amr-wind+ascent"):
+            # Necessary on Crusher to successfully find OpenMP
```
Comment (jrood-nrel) [inline var/spack/repos/builtin/packages/exawind/package.py]: Probably just remove this comment.

File: var/spack/repos/builtin/packages/nalu-wind/package.py [hunk 4]
```diff
@@ -78,12 +92,42 @@ class NaluWind(CMakePackage, CudaPackage):
...
+            when="+hypre+rocm amdgpu_target={0}".format(_arch),
+        )
+
     depends_on("trilinos-catalyst-ioss-adapter", when="+catalyst")
     depends_on("fftw+mpi", when="+fftw")
     depends_on("nccmp")
     # indirect dependency needed to make original concretizer work
     depends_on("netcdf-c+parallel-netcdf")
     depends_on("boost +filesystem +iostreams cxxstd=14", when="+boost")
+    cxxstd = ["14", "17"]
```
Comment (jrood-nrel) [inline var/spack/repos/builtin/packages/nalu-wind/package.py]: Should we abandon c++14?

File: var/spack/repos/builtin/packages/exawind/package.py (hunk 1 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/exawind/package.py]: I configured this so that `+sycl` is `amr-wind~hypre+sycl` and `nalu-wind+hypre`  with nalu-wind on host.

File: var/spack/repos/builtin/packages/nalu-wind/package.py (hunk 4 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/nalu-wind/package.py]: oh yeah probably. it won't build without c++17 now

File: var/spack/repos/builtin/packages/exawind/package.py (hunk 2 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/exawind/package.py]: sure thing

File: var/spack/repos/builtin/packages/exawind/package.py (hunk 1 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/exawind/package.py]: This did need to go though. I added some more sycl clean up.

**Review by balay — DISMISSED** [review]
hypre change looks fine to me - but good if @osborn9 can also check and approve

File: var/spack/repos/builtin/packages/exawind/package.py (hunk 1 above)
Comment (balay) [reply var/spack/repos/builtin/packages/exawind/package.py]: > hypre can build with +sycl so I think we can remove this.

Last I tried - `hypre+sycl` didn't work for me.

https://github.com/spack/spack/pull/40722#issuecomment-1792966471

Comment (jrood-nrel) [reply var/spack/repos/builtin/packages/exawind/package.py]: It may have very well gotten broken, and I haven't tried the latest Spack, but I use it successfully on the Aurora machine.

**Review by eugeneswalker — DISMISSED** [review]

**Review by osborn9 — APPROVED** [review]
The new variant in the hypre package looks good.

File: var/spack/repos/builtin/packages/amr-wind/package.py [hunk 5]
```diff
@@ -86,6 +98,16 @@ def cmake_args(self):
...
 
+        if "+hdf5" in spec:
+            cmake_options.append(self.define("AMR_WIND_ENABLE_HDF5", True))
+            cmake_options.append(self.define("AMR_WIND_ENABLE_HDF5_ZFP", True))
+            # Help AMReX understand if HDF5 is parallel or not.
+            # Building HDF5 with CMake as Spack does, causes this inspection to break.
+            if "+mpi" in spec:
+                cmake_options.append(self.define("HDF5_IS_PARALLEL", True))
+            else:
+                cmake_options.append(self.define("HDF5_IS_PARALLEL", False))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/amr-wind/package.py]:
```suggestion
            cmake_options.append(self.define("HDF5_IS_PARALLEL", spec.satisfies("+mpi")))
```

File: var/spack/repos/builtin/packages/amr-wind/package.py [hunk 6]
```diff
@@ -29,9 +29,21 @@ class AmrWind(CMakePackage, CudaPackage, ROCmPackage):
...
+    variant("umpire", default=False, description="Enable Umpire")
+    variant("sycl", default=False, description="Enable SYCL backend")
+    variant("gpu-aware-mpi", default=False, description="gpu-aware-mpi")
 
     depends_on("hypre~int64@2.20.0:", when="+hypre")
     depends_on("hypre+mpi", when="+hypre+mpi")
+    depends_on("hdf5~mpi", when="+hdf5~mpi")
+    depends_on("hdf5+mpi", when="+hdf5+mpi")
+    depends_on("h5z-zfp", when="+hdf5")
+    depends_on("zfp", when="+hdf5")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/amr-wind/package.py:45]: (Comment, not request) These can be grouped under `with when("+hdf5"):`  if it makes sense

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 7]
```diff
@@ -6,35 +6,93 @@
...
 
-class Exawind(CMakePackage):
+class Exawind(CMakePackage, CudaPackage, ROCmPackage):
     """Multi-application driver for Exawind project."""
 
     homepage = "https://github.com/Exawind/exawind-driver"
     git = "https://github.com/Exawind/exawind-driver.git"
 
-    maintainers("jrood-nrel", "psakievich")
+    maintainers = ["jrood-nrel"]
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/exawind/package.py]: Can you use the directive:
```suggestion
    maintainers("jrood-nrel")
```
?

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 8]
```diff
@@ -6,35 +6,93 @@
...
+        depends_on("nalu-wind+rocm amdgpu_target=%s" % arch, when="+rocm amdgpu_target=%s" % arch)
+
+    depends_on("nalu-wind+tioga")
+    depends_on("amr-wind+netcdf+mpi")
+    depends_on("tioga~nodegid")
     depends_on("yaml-cpp@0.6:")
+    depends_on("nalu-wind+openfast", when="+openfast")
+    depends_on("openfast+cxx@2.6.0:", when="+openfast")
+    depends_on("openfast+cxx@2.6.0:", when="^nalu-wind+openfast")
+    depends_on("openfast+cxx@2.6.0:", when="^amr-wind+openfast")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/exawind/package.py]: Hmm, not sure about this. Why do we need an explicit dependency based on how another dependency is configured? Shouldn't this go through the `openfast` variant?

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 9]
```diff
@@ -6,35 +6,93 @@
...
 
-        args = [
-            self.define("Trilinos_DIR", spec["trilinos"].prefix),
-            self.define("TIOGA_DIR", spec["tioga"].prefix),
-            self.define("Nalu-Wind_DIR", spec["nalu-wind"].prefix),
-            self.define("AMR-Wind_DIR", spec["amr-wind"].prefix),
-            self.define("OpenFAST_DIR", spec["openfast"].prefix),
-            self.define("YAML-CPP_DIR", spec["yaml-cpp"].prefix),
-        ]
+        args = super(Exawind, self).cmake_args()
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/exawind/package.py]: Why is this necessary?

File: var/spack/repos/builtin/packages/nalu-wind/package.py [hunk 10]
```diff
@@ -78,12 +93,42 @@ class NaluWind(CMakePackage, CudaPackage):
...
+            when="+hypre+rocm amdgpu_target={0}".format(_arch),
+        )
+
     depends_on("trilinos-catalyst-ioss-adapter", when="+catalyst")
     depends_on("fftw+mpi", when="+fftw")
     depends_on("nccmp")
     # indirect dependency needed to make original concretizer work
     depends_on("netcdf-c+parallel-netcdf")
     depends_on("boost +filesystem +iostreams cxxstd=14", when="+boost")
+    cxxstd = ["17"]
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/nalu-wind/package.py]: Maybe use a name which is less likely to clash (this will be defined at the class level)

**Review by alalazo — CHANGES_REQUESTED** [review]
A few minor requests. Otherwise LGTM.

File: var/spack/repos/builtin/packages/exawind/package.py (hunk 9 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/exawind/package.py]: More cruft from our inheritance it looks like

File: var/spack/repos/builtin/packages/amr-wind/package.py (hunk 6 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/amr-wind/package.py:45]: Okay if we add any more then we will update this.

File: var/spack/repos/builtin/packages/nalu-wind/package.py [hunk 11]
```diff
@@ -78,12 +93,44 @@ class NaluWind(CMakePackage, CudaPackage):
...
     depends_on("nccmp")
     # indirect dependency needed to make original concretizer work
     depends_on("netcdf-c+parallel-netcdf")
     depends_on("boost +filesystem +iostreams cxxstd=14", when="+boost")
+    supported_cxxstd = ["17"]
+    variant(
+        "cxxstd", default="17", values=supported_cxxstd, multi=False, description="cxx standard"
+    )
+    variant("tests", default=True, description="Activate regression tests")
+    variant("unit-tests", default=True, description="Activate unit tests")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/nalu-wind/package.py]: These two variants are not used:
```suggestion
```

File: var/spack/repos/builtin/packages/amr-wind/package.py [hunk 12]
```diff
@@ -98,4 +117,14 @@ def cmake_args(self):
...
 
+        if "+sycl" in self.spec:
+            cmake_options.append(self.define("AMR_WIND_ENABLE_SYCL", True))
+            # SYCL GPU backend only supported with Intel's oneAPI or DPC++ compilers
+            sycl_compatible_compilers = ["dpcpp", "icpx"]
+            if not (os.path.basename(self.compiler.cxx) in sycl_compatible_compilers):
+                raise InstallError(
+                    "AMReX's SYCL GPU Backend requires DPC++ (dpcpp)"
+                    + " or the oneAPI CXX (icpx) compiler."
+                )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/amr-wind/package.py]: You can turn this into a `requires` directive like:
```python
requires(
        "%dpcpp",
        "%oneapi",
        policy="one_of",
        msg="SYCL GPU backend only supported with Intel's oneAPI or DPC++ compilers",
    )
```
The added benefit is that you'll catch errors in concretization.

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 13]
```diff
@@ -6,35 +6,88 @@
...
+            args.append(self.define("CUDAToolkit_ROOT", self.spec["cuda"].prefix))
+            args.append(self.define("EXAWIND_CUDA_ARCH", self.spec.variants["cuda_arch"].value))
+
+        if spec.satisfies("+rocm"):
+            targets = self.spec.variants["amdgpu_target"].value
+            args.append(self.define("EXAWIND_ENABLE_ROCM", True))
+            args.append("-DCMAKE_CXX_COMPILER={0}".format(self.spec["hip"].hipcc))
+            args.append("-DCMAKE_HIP_ARCHITECTURES=" + ";".join(str(x) for x in targets))
+            args.append("-DAMDGPU_TARGETS=" + ";".join(str(x) for x in targets))
+            args.append("-DGPU_TARGETS=" + ";".join(str(x) for x in targets))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/exawind/package.py]: Can you also turn this into `self.define` for consistency?

**Review by alalazo — DISMISSED** [review]
A few other minor requests.

File: var/spack/repos/builtin/packages/nalu-wind/package.py (hunk 11 above)
Comment (psakievich) [reply var/spack/repos/builtin/packages/nalu-wind/package.py]: @jrood-nrel are you okay with removing these variants? I'd like to consider it and look into adding a test function going forward that can at least run the unit tests.

Comment (jrood-nrel) [reply var/spack/repos/builtin/packages/nalu-wind/package.py]: Yes that's fine.

File: var/spack/repos/builtin/packages/exawind/package.py [hunk 14]
```diff
@@ -6,35 +6,92 @@
...
+    depends_on("amr-wind~hypre", when="~hypre")
+    depends_on("nalu-wind+hypre", when="+hypre")
+    depends_on("nalu-wind~hypre", when="~hypre")
+    depends_on("amr-wind+sycl", when="+sycl")
+    depends_on("nalu-wind+umpire", when="+umpire")
+    depends_on("amr-wind+umpire", when="+umpire")
+    depends_on("amr-wind+tiny_profile", when="+tiny_profile")
+    depends_on("nalu-wind+gpu-aware-mpi", when="+gpu-aware-mpi")
+    depends_on("amr-wind+gpu-aware-mpi", when="+gpu-aware-mpi")
+    depends_on("nalu-wind@2.0.0:", when="@1.0.0:")
```
Comment (psakievich) [inline var/spack/repos/builtin/packages/exawind/package.py:58]: I'm really torn about these versions. I don't like them so much and debated using a uniform `ecp` tag on github + an `ecp` version.  We are really snapshotting a milestone in the project more than a "release". But I like the idea of composability going forward based off the numbers. Thoughts @jrood-nrel ?

**Comment by psakievich** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/nalu-wind/package.py [hunk 15]
```diff
@@ -78,12 +98,42 @@ class NaluWind(CMakePackage, CudaPackage):
             "hypre@develop +mpi+cuda~int64~superlu-dist cuda_arch={0}".format(_arch),
             when="+hypre+cuda cuda_arch={0}".format(_arch),
         )
+    for _arch in ROCmPackage.amdgpu_targets:
+        depends_on(
+            "trilinos@13.4.0.2022.10.27: "
+            "~shared+exodus+tpetra+zoltan+stk+boost~superlu-dist~superlu"
```
Comment (jrood-nrel) [inline var/spack/repos/builtin/packages/nalu-wind/package.py]: Let's remove `+boost`.

Comment (psakievich) [reply var/spack/repos/builtin/packages/nalu-wind/package.py]:
```suggestion
            "~shared+exodus+tpetra+zoltan+stk~superlu-dist~superlu"
```

**Comment by Z0907Q** [comment]
I'm sorry to bother you.I would like to ask how long it will take to complete the update?After building for a while, we found that we couldn't use spack to build semi-automatically.I wonder if you could give me some guidance。I had a Trilinos installation error while using spack for a semi-automated installation, and I really don't know what to do about it.Hope to get your reply, thank you.

**Comment by alalazo** [comment]
@spackbot fix style

**Comment by alalazo** [comment]
If this is waiting on https://github.com/spack/spack/pull/40793#discussion_r1434720638, we can start splitting out smaller changes in other PRs. For instance, it seems changes in `openfast` can be extracted easily.

**Comment by psakievich** [comment]
@alalazo and @Z0907Q this is a floating target. @jrood-nrel and I will try to pin it down in the next week and get it mergable.

**Comment by Z0907Q** [comment]
> 这是一个浮动目标。我会在下周尝试将其固定下来并使其可合并。
Do you mean that the problem of trilinos will be solved next week?

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
Testing this locally. I'm inclined to merge and do follow-on PR's for additional issues that pop up.

**Review by eugeneswalker — APPROVED** [review]
